### PR TITLE
[New Feature] Support for installation and independent management of applications from the Helm registry

### PR DESCRIPTION
**Pull request by GLYASAI**
related issue: #921

### CODE REVIEW

File: worker/master/helmapp/store.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	lister   v1alpha1.HelmAppLister
+}
+
+// NewStorer creates a new storer.
+func NewStorer(clientset versioned.Interface,
+	resyncPeriod time.Duration,
+	workqueue workqueue.Interface,
+	finalizerQueue workqueue.Interface) Storer {
+	// create informers factory, enable and assign required informers
+	sharedInformer := externalversions.NewSharedInformerFactoryWithOptions(clientset, resyncPeriod,
```
Comment (barnettZQG) [inline worker/master/helmapp/store.go:53]: The store module is created, Is it repeated?

File: worker/master/helmapp/app.go [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+		namespace:       helmApp.GetNamespace(),
+		templateName:    helmApp.Spec.TemplateName,
+		repoName:        helmApp.Spec.AppStore.Name,
+		repoURL:         helmApp.Spec.AppStore.URL,
+		version:         helmApp.Spec.Version,
+		revision:        helmApp.Spec.Revision,
+		overrides:       helmApp.Spec.Overrides,
+		helmCmd:         helmCmd,
+		repo:            repo,
+		chartDir:        path.Join("/tmp/helm/chart", helmApp.Namespace, helmApp.Name, helmApp.Spec.Version),
```
Comment (barnettZQG) [inline worker/master/helmapp/app.go:96]: What Is chartDir?

**Review by barnettZQG — CHANGES_REQUESTED** [review]

**Comment by barnettZQG** [comment]
Question:

* Whether the number of resources used by Helm deployed applications can be entered into the existing tenant and cluster resource usage statistics.

File: api/handler/application_handler.go [hunk 3]
```diff
@@ -11,91 +12,151 @@ import (
...
+			Version:      app.Version,
+			AppStore: &v1alpha1.HelmAppStore{
+				Name: app.AppStoreName,
+				URL:  app.AppStoreURL,
+			},
+		}}
+
+	ctx1, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+	_, err := a.kubeClient.CoreV1().Namespaces().Create(ctx1, &corev1.Namespace{
```
[the remark this answers is not in the document]
Comment (GLYASAI) [reply api/handler/application_handler.go:129]: The new team does not have a namespace

File: api/handler/application_handler.go [hunk 4]
```diff
@@ -108,20 +169,55 @@ func (a *ApplicationAction) BatchCreateApp(apps *model.CreateAppRequest, tenantI
...
+	defer cancel()
+	helmApp, err := a.rainbondClient.RainbondV1alpha1().HelmApps(app.TenantID).Get(ctx, app.AppName, metav1.GetOptions{})
+	if err != nil {
+		if k8sErrors.IsNotFound(err) {
+			return errors.Wrap(bcode.ErrApplicationNotFound, "update app")
+		}
+		return errors.Wrap(err, "update app")
 	}
-	return srcApp, nil
+	if len(req.Overrides) > 0 {
```
[the remark this answers is not in the document]
Comment (GLYASAI) [reply api/handler/application_handler.go:210]: overwrites will overwrite the items in values.yaml. why should not  Overrides is []?

File: api/handler/application_handler.go [hunk 5]
```diff
@@ -241,28 +394,115 @@ func (a *ApplicationAction) checkPorts(appID string, ports []*model.AppPort) err
...
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "get app status")
+	}
+
+	var conditions []*model.AppStatusCondition
+	for _, cdt := range status.Conditions {
+		conditions = append(conditions, &model.AppStatusCondition{
+			Type:    string(cdt.Type),
+			Status:  cdt.Status,
```
[the remark this answers is not in the document]
Comment (GLYASAI) [reply api/handler/application_handler.go:412]: Rainbond app does not have conditions.

File: api/handler/service.go [hunk 6]
```diff
@@ -2036,24 +2062,21 @@ func (s *ServiceAction) isServiceClosed(serviceID string) error {
...
-	defer func() {
-		if r := recover(); r != nil {
-			logrus.Errorf("Unexpected panic occurred, rollback transaction: %v", r)
-			tx.Rollback()
+// DeleteComponentInBatch deletes components in batch.
+func (s *ServiceAction) DeleteComponentInBatch(tx *gorm.DB, components []*dbmodel.TenantServices) error {
+	for _, cpt := range components {
+		err := s.deleteComponent(tx, cpt)
+		if err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (GLYASAI) [reply api/handler/service.go:2070]: DeleteComponentInBatch is unused. I will delete it.

File: worker/master/helmapp/app.go (hunk 2 above)
Comment (GLYASAI) [reply worker/master/helmapp/app.go:96]: chartDir is the cache directory of helm charts

File: api/controller/application.go [hunk 7]
```diff
@@ -21,14 +23,33 @@ func (a *ApplicationController) CreateApp(w http.ResponseWriter, r *http.Request
...
+	if tenantReq.AppType == model.AppTypeHelm {
+		if tenantReq.AppStoreName == "" {
+			httputil.ReturnBcodeError(r, w, bcode.NewBadRequest("the field 'app_tore_name' is required"))
+			return
+		}
+		if tenantReq.AppTemplateName == "" {
+			httputil.ReturnBcodeError(r, w, bcode.NewBadRequest("the field 'app_template_name' is required"))
+			return
+		}
+		if tenantReq.AppTemplateName == "" {
```
[the remark this answers is not in the document]
Comment (GLYASAI) [reply api/controller/application.go:35]: fixed

File: api/handler/application_handler.go [hunk 8]
```diff
@@ -241,28 +394,115 @@ func (a *ApplicationAction) checkPorts(appID string, ports []*model.AppPort) err
...
+	var services []*model.AppService
+	for _, service := range appServices.Services {
+		svc := &model.AppService{
+			ServiceName: service.Name,
+			Address:     service.Address,
+		}
+
+		svc.Pods = a.convertPods(service.Pods)
+		svc.OldPods = a.convertPods(service.OldPods)
+		svc.TCPPorts = append(svc.TCPPorts, service.TcpPorts...)
```
[the remark this answers is not in the document]
Comment (GLYASAI) [reply api/handler/application_handler.go:485]: The internal dependency only supports TCP, so the udp ports in the service are not recognized

File: api/handler/service.go [hunk 9]
```diff
@@ -750,9 +751,34 @@ func (s *ServiceAction) ServiceCreate(sc *api_model.ServiceStruct) error {
 		return err
 	}
 	logrus.Debugf("create a new app %s success", ts.ServiceAlias)
+
+	if sc.Kind == dbmodel.ServiceKindThirdParty.String() {
```
[the remark this answers is not in the document]
Comment (GLYASAI) [reply api/handler/service.go:755]: ok. i will fix it.

**Comment by GLYASAI** [comment]
> Question:
> 
> * Whether the number of resources used by Helm deployed applications can be entered into the existing tenant and cluster resource usage statistics.

The statistics of the resource of the helm application are work:
1. tenant statistics based on namespace.
2. cluster statistics based on namespace.

**Pull request merged**